crates/paimon/src/spec/types.rs

```diff
@@ -0,0 +1,985 @@
+            }
+            Self::Tinyint => {
+                DataTypeFamily::PREDEFINED
+                    | DataTypeFamily::NUMERIC
+                    | DataTypeFamily::INTEGER_NUMERIC
+                    | DataTypeFamily::EXACT_NUMERIC
+            }
+            Self::Smallint => {
+                DataTypeFamily::PREDEFINED
+                    | DataTypeFamily::NUMERIC
+                    | DataTypeFamily::INTEGER_NUMERIC
+                    | DataTypeFamily::EXACT_NUMERIC
+            }
+            Self::Integer => {
+                DataTypeFamily::PREDEFINED
+                    | DataTypeFamily::NUMERIC
+                    | DataTypeFamily::INTEGER_NUMERIC
+                    | DataTypeFamily::EXACT_NUMERIC
+            }
+            Self::Bigint => {
+                DataTypeFamily::PREDEFINED
+                    | DataTypeFamily::NUMERIC
+                    | DataTypeFamily::INTEGER_NUMERIC
+                    | DataTypeFamily::EXACT_NUMERIC
+            }
+            Self::Float => {
+                DataTypeFamily::PREDEFINED
+                    | DataTypeFamily::NUMERIC
+                    | DataTypeFamily::APPROXIMATE_NUMERIC
+            }
+            Self::Double => {
+                DataTypeFamily::PREDEFINED
+                    | DataTypeFamily::NUMERIC
+                    | DataTypeFamily::APPROXIMATE_NUMERIC
+            }
+            Self::Date => DataTypeFamily::PREDEFINED | DataTypeFamily::DATETIME,
+            Self::TimeWithoutTimeZone => {
+                DataTypeFamily::PREDEFINED | DataTypeFamily::DATETIME | DataTypeFamily::TIME
+            }
+            Self::TimestampWithoutTimeZone => {
+                DataTypeFamily::PREDEFINED | DataTypeFamily::DATETIME | DataTypeFamily::TIMESTAMP
+            }
+            Self::TimestampWithLocalTimeZone => {
+                DataTypeFamily::PREDEFINED
+                    | DataTypeFamily::DATETIME
+                    | DataTypeFamily::TIMESTAMP
+                    | DataTypeFamily::EXTENSION
+            }
+            Self::Array => DataTypeFamily::CONSTRUCTED | DataTypeFamily::COLLECTION,
+            Self::Multiset => DataTypeFamily::CONSTRUCTED | DataTypeFamily::COLLECTION,
+            Self::Map => DataTypeFamily::CONSTRUCTED | DataTypeFamily::EXTENSION,
+            Self::Row => DataTypeFamily::CONSTRUCTED,
+        }
+    }
+}
+
+/// A visitor that can visit different data types.
+pub trait DataTypeVisitor<R> {
+    fn visit(&mut self, data_type: &DataType) -> R;
+}
+
+/// Data type for paimon table.
+///
+/// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L45>
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Hash)]
+pub struct DataType {
+    is_nullable: bool,
+    type_root: DataTypeRoot,
+}
+
+impl Display for DataType {
```

Comment:
I think we should impl Display for the concrete types such as `BigIntType` not for the `DataType`?